src/herder/TxSetUtils.cpp

```diff
@@ -0,0 +1,399 @@
+                            "result: {}",
+                            hexAbbrev(txSet.previousLedgerHash()), lastSeq,
+                            xdr_to_string(tx->getEnvelope(),
+                                          "TransactionEnvelope"),
+                            tx->getResultCode());
+                    }
+                    return invalidTxs;
+                }
+            }
+            else // update the account fee map
+            {
+                lastSeq = tx->getSeqNum();
+                int64_t& accFee = accountFeeMap[tx->getFeeSourceID()];
+                if (INT64_MAX - accFee < tx->getFeeBid())
+                {
+                    accFee = INT64_MAX;
+                }
+                else
+                {
+                    accFee += tx->getFeeBid();
+                }
+                ++iter;
+            }
+        }
+    }
+
+    auto header = ltx.loadHeader();
+    for (auto& kv : accountTxMap)
+    {
+        auto iter = kv.second.begin();
+        while (iter != kv.second.end())
+        {
+            auto tx = *iter;
+            auto feeSource = stellar::loadAccount(ltx, tx->getFeeSourceID());
+            auto totFee = accountFeeMap[tx->getFeeSourceID()];
+            if (getAvailableBalance(header, feeSource) < totFee)
+            {
+                while (iter != kv.second.end())
+                {
+                    invalidTxs.emplace_back(*iter);
+                    ++iter;
+                }
+                if (returnEarlyOnFirstInvalidTx)
+                {
+                    CLOG_DEBUG(Herder,
+                               "Got bad txSet: {} account can't pay fee tx: {}",
+                               hexAbbrev(txSet.previousLedgerHash()),
+                               xdr_to_string(tx->getEnvelope(),
+                                             "TransactionEnvelope"));
+                    return invalidTxs;
+                }
+            }
+            else
+            {
+                ++iter;
+            }
+        }
+    }
+
+    return invalidTxs;
+}
+
+// Target use case is to remove a subset of invalid transactions from a TxSet.
+// I.e. txSet.size() >= txsToRemove.size()
+TxSetFrameConstPtr
+removeTxs(TxSetFrameConstPtr txSet, TxSetFrame::Transactions const& txsToRemove)
+{
+    // hashmap from the full txSet
+    std::unordered_map<Hash, TransactionFrameBasePtr> fullTxSetHashMap;
+    fullTxSetHashMap.reserve(txSet->sizeTx());
+    std::transform(txSet->getTxsInHashOrder().cbegin(),
+                   txSet->getTxsInHashOrder().cend(),
+                   std::inserter(fullTxSetHashMap, fullTxSetHashMap.end()),
+                   [](TransactionFrameBasePtr const& tx) {
+                       return std::pair<Hash, TransactionFrameBasePtr>{
+                           tx->getFullHash(), tx};
+                   });
+
+    // candidate tx hashes to be removed
+    std::unordered_set<Hash> removeTxHashSet;
+    removeTxHashSet.reserve(txsToRemove.size());
```

Comment:
Yeah I must've been thinking some weird cases.. Cleaned it up.

Reply:
I didn't see you were looping again at the end... so a better algo (which also will work better with GeneralizedTransactions groups) is:
* create an `UnorderedSet` for `txsToRemove` (note: we use that instead of `std::unordered_set` as much as possible)
* iterate over the input txset and construct `txs` by excluding transactions that are found in `txsToRemove`

less copying, less code.

Reply:
Fixed. Yeah I agree this is better.